core/drivers/atmel_shdwc_a32.S

```diff
@@ -0,0 +1,59 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2015 Atmel Corporation,
+ *                    Nicolas Ferre <nicolas.ferre@atmel.com>
+ * Copyright (c) 2021, Microchip
+ */
+
+
+#include <arm.h>
+#include <arm32_macros.S>
+#include <asm.S>
+#include <at91_pmc.h>
+#include <drivers/atmel_shdwc.h>
+#include <drivers/sam/at91_ddr.h>
+
+/**
+ * Shutdown the CPU
+ *
+ * This function is in assembly to be sure the code fits in a single cache line.
+ * We are going to power down the SDRAM and thus we can't fetch code from it
+ * once powered down.
+ *
+ * r0 = mpddrc_base
+ * r1 = shdwc_base
+ * r2 = pmc_base
+ */
+FUNC atmel_shdwc_shutdown_asm , :
+
+	ldr r3, =AT91_DDRSDRC_LPDDR2_PWOFF
+	ldr r4, =(AT91_SHDW_KEY | AT91_SHDW_SHDW)
+
+/* Align to cache line */
+.balign 32
+	/* Ensure AT91_SHDW_CR is in the TLB by reading it */
```

Comment:
For how long do you expect it will stay here?
Any chance that it can get evicted earlier than expected?

Reply:
If no other access are made after that, the entry should still be in the TLB after that since the micro TLB is 10 entries wide.
So at least, PMC, MPDDRC, SHDWC and current code translation will fit in it.

To be sure of that, I will also disable exception before entering this piece of code so nothing will happen in this path. I also noticed that with this code, the PMC is not accessed before powering down the SDRAM but this does not affect the sama5d2_xplained because we don't have LPDDR on it.

Reply:
What if another CPU invalidates the TLB?
I guess this might become more of a DOS than a security issue.